File: src/ts/fuzz.ts
// Deterministic random session generation — the TypeScript mirror of
// src/rust/src/fuzz.rs (same xorshift64*, same vocabulary, same probabilities)
// so every runtime fuzzes itself with identical sequences.
import { ORES_DND_PROTOCOL, type DndEnvelope, type DndItem, type DndItemKind, type DndOperation } from "./codec.js";
import type { DndDropPolicy } from "./policy.js";
import { inputs, type DndSessionInput } from "./session.js";

const MASK64 = (1n << 64n) - 1n;

/** xorshift64* over BigInt; reproduces the Rust generator bit for bit. */
export class Xorshift64 {
  #state: bigint;

  constructor(seed: bigint | number) {
    const s = BigInt(seed) & MASK64;
    this.#state = s === 0n ? 0x9e3779b97f4a7c15n : s;
  }

  nextU64(): bigint {
    let x = this.#state;
    x ^= x >> 12n;
    x ^= (x << 25n) & MASK64;
    x ^= x >> 27n;
    this.#state = x;
    return (x * 0x2545f4914f6cdd1dn) & MASK64;
  }

  below(n: number): number {
    return Number((this.nextU64() >> 32n) % BigInt(n));
  }

  chance(percent: number): boolean {
    return this.below(100) < percent;
  }

  pick<T>(items: readonly T[]): T {
    return items[this.below(items.length)]!;
  }
}

export const OPS: readonly DndOperation[] = ["copy", "move", "link"];
export const KINDS: readonly DndItemKind[] = ["text", "uri", "json", "bytes"];
export const MEDIA = ["text/plain", "text/markdown", "text/uri-list", "application/json", "image/png"] as const;
export const MEDIA_PATTERNS = ["text/plain", "text/*", "application/json", "image/*", "text/markdown", "application/*"] as const;
export const TARGETS = ["zone-a", "zone-b", "zone-c", "zone-d"] as const;
export const FORMS = ["form-x", "form-y"] as const;

function opsSubset(rng: Xorshift64): DndOperation[] {
  const ops = OPS.filter(() => rng.chance(55));
  if (ops.length === 0) ops.push(rng.pick(OPS));
  return ops;
}

export function randomEnvelope(rng: Xorshift64, n: number): DndEnvelope {
  const itemCount = 1 + rng.below(3);
  const items: DndItem[] = [];
  for (let i = 0; i < itemCount; i += 1) {
    const media = rng.pick(MEDIA);
    const kind: DndItemKind = media === "text/uri-list" ? "uri" : media === "application/json" ? "json" : media === "image/png" ? "bytes" : "text";
    items.push({ kind, mediaType: media, data: "x".repeat(rng.below(9)) });
  }
  const allowedOperations = opsSubset(rng);
  const envelope: DndEnvelope = {
    protocol: rng.chance(8) ? "ores.dnd/v2" : ORES_DND_PROTOCOL,
    dragId: `drag-${String(n).padStart(4, "0")}`,
    sourceRuntime: "fuzz",
    allowedOperations,
    items,
  };
  if (rng.chance(30)) envelope.formId = rng.pick(FORMS);
  return envelope;
}

export function randomPolicy(rng: Xorshift64): DndDropPolicy {
  const kinds = KINDS.filter(() => rng.chance(55));
  if (kinds.length === 0) kinds.push(rng.pick(KINDS));
  const policy: DndDropPolicy = { targetId: rng.pick(TARGETS), allowedOperations: opsSubset(rng), acceptedKinds: kinds };
  if (rng.chance(35)) {
    const patterns = MEDIA_PATTERNS.filter(() => rng.chance(40)).map(String);
    if (patterns.length === 0) patterns.push(rng.pick(MEDIA_PATTERNS));
    policy.acceptedMediaTypes = patterns;
  }
  if (rng.chance(30)) policy.maxItems = 1 + rng.below(3);
  if (rng.chance(30)) policy.maxTotalBytes = 1 + rng.below(16);
  if (rng.chance(25)) policy.formId = rng.pick(FORMS);
  return policy;
}

export function randomInput(rng: Xorshift64, counter: { n: number }): DndSessionInput {
  const roll = rng.below(100);
  if (roll <= 17) {
    counter.n += 1;
    return inputs.start(randomEnvelope(rng, counter.n));
  }
  if (roll <= 52) {
    const policy = randomPolicy(rng);
    const preferred = rng.chance(30) ? rng.pick(OPS) : undefined;
    return inputs.enter(policy, preferred);
  }
  if (roll <= 67) return inputs.leave(rng.pick(TARGETS));
  if (roll <= 85) return inputs.drop(rng.pick(TARGETS));
  if (roll <= 92) return inputs.cancel();
  return inputs.end();
}

/** A full random sequence, always beginning with a valid start. */
export function randomSequence(seed: bigint | number, steps: number): DndSessionInput[] {
  const rng = new Xorshift64(seed);
  const counter = { n: 1 };
  const first = randomEnvelope(rng, counter.n);
  first.protocol = ORES_DND_PROTOCOL;
  const out: DndSessionInput[] = [inputs.start(first)];
  while (out.length < steps) out.push(randomInput(rng, counter));
  return out;
}

File: src/rust/src/fuzz.rs
//! Deterministic random session generation shared by the fuzz-trace generator
//! (`examples/gen_fuzz_traces.rs`) and the invariant tests. A tiny xorshift
//! PRNG keeps the corpus reproducible without a `rand` dependency; the same
//! generator is mirrored in TypeScript and Dart so every runtime can fuzz
//! itself with identical sequences.

use crate::envelope::{DndEnvelope, DndItem, DndItemKind, DndOperation, ORES_DND_PROTOCOL};
use crate::policy::DndDropPolicy;
use crate::session::DndSessionInput;

/// xorshift64* — reproducible across languages (64-bit wrapping arithmetic).
#[derive(Debug, Clone)]
pub struct Xorshift64(pub u64);

impl Xorshift64 {
    pub fn new(seed: u64) -> Self {
        Self(if seed == 0 { 0x9E37_79B9_7F4A_7C15 } else { seed })
    }

    pub fn next_u64(&mut self) -> u64 {
        let mut x = self.0;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.0 = x;
        x.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }

    /// Uniform in `0..n` (n > 0); uses the high bits for quality.
    pub fn below(&mut self, n: u32) -> u32 {
        ((self.next_u64() >> 32) % u64::from(n)) as u32
    }

    pub fn chance(&mut self, percent: u32) -> bool {
        self.below(100) < percent
    }

    pub fn pick<'a, T>(&mut self, items: &'a [T]) -> &'a T {
        &items[self.below(items.len() as u32) as usize]
    }
}

pub const OPS: [DndOperation; 3] = [DndOperation::Copy, DndOperation::Move, DndOperation::Link];
pub const KINDS: [DndItemKind; 4] = [DndItemKind::Text, DndItemKind::Uri, DndItemKind::Json, DndItemKind::Bytes];
pub const MEDIA: [&str; 5] = ["text/plain", "text/markdown", "text/uri-list", "application/json", "image/png"];
pub const MEDIA_PATTERNS: [&str; 6] = ["text/plain", "text/*", "application/json", "image/*", "text/markdown", "application/*"];
pub const TARGETS: [&str; 4] = ["zone-a", "zone-b", "zone-c", "zone-d"];
pub const FORMS: [&str; 2] = ["form-x", "form-y"];

pub fn random_envelope(rng: &mut Xorshift64, n: u32) -> DndEnvelope {
    let item_count = 1 + rng.below(3) as usize;
    let items = (0..item_count)
        .map(|_| {
            let media = *rng.pick(&MEDIA);
            let kind = match media {
                "text/uri-list" => DndItemKind::Uri,
                "application/json" => DndItemKind::Json,
                "image/png" => DndItemKind::Bytes,
                _ => DndItemKind::Text,
            };
            DndItem { kind, media_type: media.to_owned(), data: "x".repeat(rng.below(9) as usize), name: None }
        })
        .collect();
    let ops = ops_subset(rng);
    DndEnvelope {
        protocol: if rng.chance(8) { "ores.dnd/v2".to_owned() } else { ORES_DND_PROTOCOL.to_owned() },
        drag_id: format!("drag-{n:04}"),
        source_runtime: "fuzz".to_owned(),
        allowed_operations: ops,
        items,
        traceparent: None,
        form_id: if rng.chance(30) { Some((*rng.pick(&FORMS)).to_owned()) } else { None },
    }
}

fn ops_subset(rng: &mut Xorshift64) -> Vec<DndOperation> {
    let mut ops: Vec<DndOperation> = OPS.iter().copied().filter(|_| rng.chance(55)).collect();
    if ops.is_empty() {
        ops.push(*rng.pick(&OPS));
    }
    ops
}

pub fn random_policy(rng: &mut Xorshift64) -> DndDropPolicy {
    let mut kinds: Vec<DndItemKind> = KINDS.iter().copied().filter(|_| rng.chance(55)).collect();
    if kinds.is_empty() {
        kinds.push(*rng.pick(&KINDS));
    }
    let mut policy = DndDropPolicy::new(*rng.pick(&TARGETS), &ops_subset(rng), &kinds);
    if rng.chance(35) {
        let mut patterns: Vec<String> = MEDIA_PATTERNS.iter().filter(|_| rng.chance(40)).map(|p| (*p).to_owned()).collect();
        if patterns.is_empty() {
            patterns.push((*rng.pick(&MEDIA_PATTERNS)).to_owned());
        }
        policy = policy.with_media_types(patterns);
    }
    if rng.chance(30) {
        policy = policy.with_max_items(1 + rng.below(3) as i32);
    }
    if rng.chance(30) {
        policy = policy.with_max_total_bytes(1 + rng.below(16) as i32);
    }
    if rng.chance(25) {
        policy = policy.with_form_id(*rng.pick(&FORMS));
    }
    policy
}

/// One random input; `n` numbers envelopes so drag ids stay unique per trace.
pub fn random_input(rng: &mut Xorshift64, n: &mut u32) -> DndSessionInput {
    match rng.below(100) {
        0..=17 => {
            *n += 1;
            DndSessionInput::start(random_envelope(rng, *n))
        }
        18..=52 => {
            let policy = random_policy(rng);
            let preferred = if rng.chance(30) { Some(*rng.pick(&OPS)) } else { None };
            DndSessionInput::enter(policy, preferred)
        }
        53..=67 => DndSessionInput::leave(*rng.pick(&TARGETS)),
        68..=85 => DndSessionInput::drop(*rng.pick(&TARGETS)),
        86..=92 => DndSessionInput::cancel(),
        _ => DndSessionInput::end(),
    }
}

/// A full random sequence, always beginning with a valid `start`.
pub fn random_sequence(seed: u64, steps: usize) -> Vec<DndSessionInput> {
    let mut rng = Xorshift64::new(seed);
    let mut n = 0;
    let mut inputs = Vec::with_capacity(steps);
    n += 1;
    let mut first = random_envelope(&mut rng, n);
    first.protocol = ORES_DND_PROTOCOL.to_owned();
    inputs.push(DndSessionInput::start(first));
    while inputs.len() < steps {
        inputs.push(random_input(&mut rng, &mut n));
    }
    inputs
}
